File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/metadata/StorageEntityId.java
package com.foundgine.core.semantic.metadata;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.KeyDeserializer;
import com.fasterxml.jackson.databind.SerializerProvider;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.fasterxml.jackson.databind.deser.std.StdDeserializer;
import com.fasterxml.jackson.databind.ser.std.StdSerializer;
import com.foundgine.core.abstractions.SemanticIdentity;
import com.foundgine.core.abstractions.UnsignedLongJson;
import java.io.IOException;

/** Identity of a physical storage entity such as a database table. */
@JsonSerialize(using = StorageEntityId.Serializer.class, keyUsing = StorageEntityId.KeySerializer.class)
@JsonDeserialize(using = StorageEntityId.Deserializer.class, keyUsing = StorageEntityId.KeyDeserializer.class)
public record StorageEntityId(long value) {
    public static StorageEntityId create(String storageName) {
        return new StorageEntityId(SemanticIdentity.hash(SemanticIdentity.tableKey(storageName)));
    }
    public static final class Serializer extends StdSerializer<StorageEntityId> {
        public Serializer() { super(StorageEntityId.class); }
        @Override public void serialize(StorageEntityId id, JsonGenerator g, SerializerProvider p) throws IOException {
            UnsignedLongJson.writeNumber(g, id.value());
        }
    }
    public static final class Deserializer extends StdDeserializer<StorageEntityId> {
        public Deserializer() { super(StorageEntityId.class); }
        @Override public StorageEntityId deserialize(JsonParser p, DeserializationContext c) throws IOException {
            return new StorageEntityId(UnsignedLongJson.read(p, "StorageEntityId"));
        }
    }
    public static final class KeySerializer extends StdSerializer<StorageEntityId> {
        public KeySerializer() { super(StorageEntityId.class); }
        @Override public void serialize(StorageEntityId id, JsonGenerator g, SerializerProvider p) throws IOException {
            UnsignedLongJson.writeFieldName(g, id.value());
        }
    }
    public static final class KeyDeserializer extends KeyDeserializer {
        @Override public StorageEntityId deserializeKey(String key, DeserializationContext c) {
            return new StorageEntityId(UnsignedLongJson.parseUnsigned(key));
        }
    }
}

File: src/csharp/Foundgine.Core/Semantic/Metadata/StorageEntityId.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic.Metadata;

/// <summary>Identity of a physical storage entity such as a database table.</summary>
[JsonConverter(typeof(StorageEntityIdJsonConverter))]
public readonly record struct StorageEntityId(ulong Value)
{
    public static StorageEntityId Create(string storageName) =>
        new(SemanticIdentity.Hash(SemanticIdentity.TableKey(storageName)));
}

public sealed class StorageEntityIdJsonConverter : JsonConverter<StorageEntityId>
{
    public override StorageEntityId Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        if (reader.TokenType == JsonTokenType.Number)
            return new StorageEntityId(reader.GetUInt64());
        if (reader.TokenType == JsonTokenType.StartObject)
        {
            using var document = JsonDocument.ParseValue(ref reader);
            if (document.RootElement.TryGetProperty("Value", out var value))
                return new StorageEntityId(value.GetUInt64());
        }

        throw new JsonException("Expected a StorageEntityId numeric value or a legacy {\"Value\":...} object.");
    }

    public override void Write(Utf8JsonWriter writer, StorageEntityId value, JsonSerializerOptions options) =>
        writer.WriteNumberValue(value.Value);

    public override StorageEntityId ReadAsPropertyName(ref Utf8JsonReader reader, Type typeToConvert,
        JsonSerializerOptions options) => new(ulong.Parse(reader.GetString()!));

    public override void WriteAsPropertyName(Utf8JsonWriter writer, StorageEntityId value,
        JsonSerializerOptions options) =>
        writer.WritePropertyName(value.Value.ToString(System.Globalization.CultureInfo.InvariantCulture));
}